File: java/src/main/java/org/katas/CurrencyConverter.java
package org.katas;

import org.katas.external.ConversionRateApi;
import org.katas.model.Currency;
import org.katas.model.Money;

import java.util.HashMap;

public class CurrencyConverter {
    private final ConversionRateApi rateApi;

    CurrencyConverter(ConversionRateApi rateApi) {
        this.rateApi = rateApi;
    }

    Money sum(Currency target, Money... moneys) {
        // We are paying each call we make to the API, so better not make useless calls
        double amount = 0.0;

        var knownRates = new HashMap<Currency, Double>();
        knownRates.put(target, 1.0);

        for (var m: moneys) {
            if (!knownRates.containsKey(m.currency())){
                knownRates.put(m.currency(), this.rateApi.getRate(m.currency().toIsoCode(), target.toIsoCode()));
            }
            amount += m.amount() * knownRates.get(m.currency());
        }

        return new Money(amount, target);
    }
}

File: kotlin/src/main/kotlin/org/katas/CurrencyConverter.kt
package org.katas

import org.katas.external.conversionRateApi.ConversionRateApi
import org.katas.model.Currency
import org.katas.model.Money
import java.util.*

class CurrencyConverter(private val rateApi: ConversionRateApi) {
    fun sum(target: Currency, vararg moneys: Money): Money {
        // We are paying each call we make to the API, so better not make useless calls
        var amount = 0.0
        val knownRates = HashMap<Currency, Double>()
        knownRates[target] = 1.0
        for (m in moneys) {
            if (!knownRates.containsKey(m.currency)) {
                knownRates[m.currency] = rateApi.getRate(m.currency.toIsoCode(), target.toIsoCode())
            }
            amount += m.amount * knownRates[m.currency]!!
        }
        return Money(amount, target)
    }
}